File: swift/Sources/LoomupAppIntegrity/MobileClient.swift
import Foundation
import Loomup

/// Secure-by-default iOS construction: App Attest for mutations, refresh token
/// in Keychain, and access token only in memory. There is intentionally no
/// service-key parameter.
@available(iOS 16.0, *)
public func createMobileClient(
    url: URL,
    appID: String,
    token: String? = nil,
    http: any HTTPTransport = URLSessionHTTPTransport(),
    webSocketFactory: WebSocketFactory? = nil
) -> LoomupClient {
    let refreshStore = KeychainRefreshTokenStore(account: appID)
    let integrity = AppleAppIntegrityProvider(baseURL: url, appID: appID, http: http)
    return LoomupClient(
        options: LoomupClientOptions(
            url: url,
            token: token,
            appIntegrityProvider: integrity,
            refreshTokenStore: refreshStore,
            http: http,
            webSocketFactory: webSocketFactory
        )
    )
}

File: kotlin/android/src/main/kotlin/com/loomup/client/android/MobileClient.kt
package com.loomup.client.android

import android.content.Context
import com.loomup.client.HttpTransport
import com.loomup.client.LoomupClient
import com.loomup.client.LoomupClientOptions
import com.loomup.client.OkHttpHttpTransport
import com.loomup.client.WebSocketFactory

/** Secure mobile construction with no service-key input. */
fun createAndroidClient(
    context: Context,
    url: String,
    appId: String,
    cloudProjectNumber: Long,
    token: String? = null,
    http: HttpTransport = OkHttpHttpTransport(),
    webSocketFactory: WebSocketFactory? = null,
): LoomupClient {
    val refreshStore = AndroidRefreshTokenStore(context, appId)
    val integrity = PlayIntegrityProvider(context, url, appId, cloudProjectNumber, http)
    return LoomupClient(
        LoomupClientOptions(
            url = url,
            token = token,
            appIntegrityProvider = integrity,
            refreshTokenStore = refreshStore,
            http = http,
            webSocketFactory = webSocketFactory,
        ),
    )
}
